network/endpoint_linux.go

```diff
@@ -54,204 +55,203 @@ func (nw *network) newEndpointImpl(
 	}
 
-	if epInfo.Data != nil {
-		if _, ok := epInfo.Data[VlanIDKey]; ok {
-			vlanid = epInfo.Data[VlanIDKey].(int)
-		}
-
-		if _, ok := epInfo.Data[LocalIPKey]; ok {
-			localIP = epInfo.Data[LocalIPKey].(string)
-		}
-	}
+	for _, epInfo := range epInfo {
+		epClient := epClient
+		switch epInfo.AddressType {
+		case cns.Secondary:
+			log.Printf("Secondary client")
+			if ep.SecondaryInterfaces == nil {
+				ep.SecondaryInterfaces = make(map[string]*InterfaceInfo)
+			}
+			if epClient == nil {
+				epClient = NewSecondaryEndpointClient(nl, plc, ep)
+			}
+		default:
+			if nw.Endpoints[epInfo.Id] != nil {
+				log.Printf("[net] Endpoint already exists.")
+				err = errEndpointExists
+				return nil, err
+			}
 
-	if _, ok := epInfo.Data[OptVethName]; ok {
-		key := epInfo.Data[OptVethName].(string)
-		log.Printf("Generate veth name based on the key provided %v", key)
-		vethname := generateVethName(key)
-		hostIfName = fmt.Sprintf("%s%s", hostVEthInterfacePrefix, vethname)
-		contIfName = fmt.Sprintf("%s%s2", hostVEthInterfacePrefix, vethname)
-	} else {
-		// Create a veth pair.
-		log.Printf("Generate veth name based on endpoint id")
-		hostIfName = fmt.Sprintf("%s%s", hostVEthInterfacePrefix, epInfo.Id[:7])
-		contIfName = fmt.Sprintf("%s%s-2", hostVEthInterfacePrefix, epInfo.Id[:7])
-	}
+			if epInfo.Data != nil {
+				if _, ok := epInfo.Data[VlanIDKey]; ok {
+					vlanid = epInfo.Data[VlanIDKey].(int)
+				}
 
-	// epClient is non-nil only when the endpoint is created for the unit test.
-	if epClient == nil {
-		//nolint:gocritic
-		if vlanid != 0 {
-			if nw.Mode == opModeTransparentVlan {
-				log.Printf("Transparent vlan client")
-				if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
-					nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
+				if _, ok := epInfo.Data[LocalIPKey]; ok {
+					localIP = epInfo.Data[LocalIPKey].(string)
 				}
-				epClient = NewTransparentVlanEndpointClient(nw, epInfo, hostIfName, contIfName, vlanid, localIP, nl, plc)
+			}
+
+			if _, ok := epInfo.Data[OptVethName]; ok {
+				key := epInfo.Data[OptVethName].(string)
+				log.Printf("Generate veth name based on the key provided %v", key)
+				vethname := generateVethName(key)
+				hostIfName = fmt.Sprintf("%s%s", hostVEthInterfacePrefix, vethname)
+				contIfName = fmt.Sprintf("%s%s2", hostVEthInterfacePrefix, vethname)
 			} else {
-				log.Printf("OVS client")
-				if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
-					nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
-				}
+				// Create a veth pair.
+				log.Printf("Generate veth name based on endpoint id")
+				hostIfName = fmt.Sprintf("%s%s", hostVEthInterfacePrefix, epInfo.Id[:7])
+				contIfName = fmt.Sprintf("%s%s-2", hostVEthInterfacePrefix, epInfo.Id[:7])
+			}
 
-				epClient = NewOVSEndpointClient(
-					nw,
-					epInfo,
-					hostIfName,
-					contIfName,
-					vlanid,
-					localIP,
-					nl,
-					ovsctl.NewOvsctl(),
-					plc)
+			ep.Id = epInfo.Id
```

Comment:
why changing the order here (endpoint intialisation and then client init)?

Reply:
no reason, just looked cleaner

Reply:
imo the older code looks cleaner where we initialize all fields in one place instead of doing each field separately

Reply:
moved this to the top now